Sources/SemanticIndex/SemanticIndexManager.swift

```diff
@@ -233,67 +244,101 @@ public final actor SemanticIndexManager {
     return filesToReIndex
   }
 
+  /// Schedule preparation of the target that contains the given URI, building all modules that the file depends on.
+  ///
+  /// This is intended to be called when the user is interacting with the document at the given URI.
+  public func schedulePreparation(of uri: DocumentURI, priority: TaskPriority? = nil) async {
+    Task(priority: priority) {
+      await withLoggingScope("preparation") {
+        guard let target = await buildSystemManager.canonicalConfiguredTarget(for: uri) else {
+          return
+        }
+        await self.prepare(targets: [target], priority: priority)
+      }
+    }
+  }
+
   // MARK: - Helper functions
 
   /// Prepare the given targets for indexing
   private func prepare(targets: [ConfiguredTarget], priority: TaskPriority?) async {
-    let targetsToPrepare = targets.filter {
-      if case .upToDate = preparationStatus[$0] {
-        return false
+    var targetsToPrepare: [ConfiguredTarget] = []
+    var preparationTasksToAwait: [Task<Void, Never>] = []
+    for target in targets {
+      switch preparationStatus[target] {
+      case .upToDate:
+        break
+      case .scheduled((_, let configuredTargets, let task)), .executing((_, let configuredTargets, let task)):
+        // If we already have a task scheduled that prepares fewer targets, await that instead of overriding the
+        // target's preparation status with a longer-running task. The key benefit here is that when we get many
+        // preparation requests for the same target (eg. one for every text document request sent to a file), we don't
+        // re-create new `PreparationTaskDescription`s for every preparation request. Instead, all the preparation
+        // requests await the same task. At the same time, if we have a multi-file preparation request and then get a
+        // single-file preparation request, we will override the preparation of that target with the single-file
+        // preparation task, ensuring that the task gets prepared as quickly as possible.
+        if configuredTargets.count <= targets.count {
+          preparationTasksToAwait.append(task)
+        }
+        fallthrough
+      case nil:
+        targetsToPrepare.append(target)
```

Comment:
I might be misunderstanding, but won't this mean that we'd still schedule the preparation again for the same target if we already have one running?

Reply:
Yes, it does and that’s intentional, in the context of multi-target preparation. Say that while background indexing you are preparing targets `A`, `B`, and `C` in a single SwiftPM invocation (like a hypothetical `swift prepare --target A --target B --target C`), which has the benefit that it allows concurrent building of the three targets if they don’t depend on each other. 

Now you are trying to work on a file in `A`, which requires target `A` to be prepared. Now, you don’t want to wait for preparation of all three targets to finish, but want to build target `A` as quickly as possible. So, you schedule preparation of `A` with high priority, which will cancel and reschedule the preparation of `A`, `B` and `C` in `TaskScheduler`. Once preparation of `A` has finished, the joint preparation of `A`, `B`, and `C` is re-scheduled, which will effectively only need to prepare `B` and `C` because the build system realizes that `A` is already up-to-date.

Does that make sense?

Reply:
Yeah that makes sense. I was mainly thinking of the case where you have `A` being prepared and get a new request to prepare `A`, we'll schedule it again. Though I wasn't really thinking about the scheduling logic, now that I think about it, we'll make the subsequent request wait on the former, and then when evaluating it should be a no-op, so I guess this isn't the end of the world.

Reply:
This doc comment doesn't quite seem true though:

```
        // target's preparation status with a longer-running task. The key benefit here is that when we get many
        // preparation requests for the same target (eg. one for every text document request sent to a file), we don't
        // re-create new `PreparationTaskDescription`s for every preparation request. 
```

Reply:
Ah, I was just missing a `break` here.